server/interaction/npc.go

```diff
@@ -0,0 +1,123 @@
+package interaction
+
+import (
+	"context"
+	"errors"
+
+	"github.com/joshprzybyszewski/cribbage/logic/scorer"
+	"github.com/joshprzybyszewski/cribbage/model"
+	"github.com/joshprzybyszewski/cribbage/utils/rand"
+)
+
+const (
+	Dumb   model.PlayerID = `DumbNPC`
+	Simple model.PlayerID = `SimpleNPC`
+	Calc   model.PlayerID = `CalculatedNPC`
+)
+
+var (
+	ErrUnknownNPCType = errors.New(`unknown NPC type`)
+)
+
+var npcs = map[model.PlayerID]npc{
+	Dumb:   &dumbNPC{},
+	Simple: &simpleNPC{},
+	Calc:   &calculatedNPC{},
+}
+
+var _ Player = (*NPCPlayer)(nil)
+
+type NPCPlayer struct {
+	HandleActionCallback func(ctx context.Context, a model.PlayerAction) error
```

Comment:
I don't usually like having funcs as struct fields because it confuses "what's a method" and "what's a field" ... hmm I wonder what we _should_ do...

Reply:
I got it. interfaces.

So the `interaction/` package needs to define an interface called `ActionHandler` and then the `server/` package needs to create an "npc action handler struct" which has that method on it. then the player means will accept an interface for the npc `interaction.Player` to use. So

In `interaction/` somewhere (probably interface.go):
```go
type ActionHandler interface {
	HandleAction(action model.PlayerAction) error
}
```

then on creation of the NewNPCPlayer in `FromPlayerMeans`:
```go
	switch means.Mode {
	case Localhost:
		return newLocalhostPlayer(pID, means.Info), nil
	case NPC:
		ah, ok := means.Info.(ActionHandler)
		if !ok { /* return an error */ }
		return NewNPCPlayer(pID, ah)
	}
```

Here in interaction/npc.go:
```go
func NewNPCPlayer(pID model.PlayerID, ah ActionHandler) (Player, error) {
```

then in `server/` (probably a new file) we make a struct to implement this interface:
```go
var _ interaction.ActionHandler = (*npcActionHandler)(nil)

type npcActionHandler struct{}

func (*npcActionHandler) HandleAction(action model.PlayerAction) error {
	return HandleAction(context.Background(), action)
}
```

and that struct could rate limit/only allow one handle action request from an npc at a time using a lock (if we wanted to):
```go
var _ interaction.ActionHandler = (*npcActionHandler)(nil)

type npcActionHandler struct {
	lock sync.Mutex
}

func (n *npcActionHandler) HandleAction(action model.PlayerAction) error {
	n.lock.Lock()
	defer n.lock.Unlock()
	return HandleAction(context.Background(), action)
}
```